plat/drivers/uart/ns16550.c

```diff
@@ -0,0 +1,193 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <libfdt.h>
+#include <uk/config.h>
+#include <uk/plat/console.h>
+#include <uk/assert.h>
+#include <arm/cpu.h>
+
+#define NS16550_THR_OFFSET	0x00U
+#define NS16550_RBR_OFFSET	0x00U
+#define NS16550_IER_OFFSET	0x01U
+#define NS16550_IIR_OFFSET	0x02U
+#define NS16550_FCR_OFFSET	0x02U
+#define NS16550_LCR_OFFSET	0x03U
+#define NS16550_MCR_OFFSET	0x04U
+#define NS16550_LSR_OFFSET	0x05U
+#define NS16550_MSR_OFFSET	0x06U
+
+#define NS16550_REG_SHIFT	0x02U
+
+#define NS16550_LCR_DLAB	0x80U
+#define NS16550_IIR_NO_INT	0x01U
+#define NS16550_FCR_FIFO_EN	0x01U
+#define NS16550_LSR_RX_EMPTY	0x01U
+#define NS16550_LSR_TX_EMPTY	0x40U
+
+/*
+ * NS16550 UART base address
+ * As we are using the PA = VA mapping, some SoC would set PA 0
+ * as a valid address, so we can't use ns16550_uart_base == 0 to
+ * indicate NS16550 hasn't been initialized. In this case, we
+ * use ns16550_uart_initialized as an extra variable to check
+ * whether the UART has been initialized.
+ */
+#if defined(CONFIG_EARLY_UART_NS16550) && defined(CONFIG_EARLY_UART_NS16550_BASE)
+static uint8_t ns16550_uart_initialized = 1;
+static uint64_t ns16550_uart_base = CONFIG_EARLY_UART_NS16550_BASE;
+#else
+static uint8_t ns16550_uart_initialized;
+static uint64_t ns16550_uart_base;
+#endif
+
+/* Macros to access ns16550 registers with base address and reg shift */
+#define NS16550_REG(r)			((uint16_t *)(ns16550_uart_base + \
+					(r << NS16550_REG_SHIFT)))
+#define NS16550_REG_READ(r)		ioreg_read16(NS16550_REG(r))
+#define NS16550_REG_WRITE(r, v)	ioreg_write16(NS16550_REG(r), v)
+
+static void init_ns16550(uint64_t base)
+{
+	ns16550_uart_base = base;
+	ns16550_uart_initialized = 1;
+
+	/* Disable all interrupts */
+	NS16550_REG_WRITE(NS16550_IER_OFFSET,
+		NS16550_REG_READ(NS16550_FCR_OFFSET) & ~(NS16550_IIR_NO_INT));
+
+	/* Disable FIFOs */
+	NS16550_REG_WRITE(NS16550_FCR_OFFSET,
+		NS16550_REG_READ(NS16550_FCR_OFFSET) & ~(NS16550_FCR_FIFO_EN));
+}
+
+void ns16550_console_init(const void *dtb)
+{
+	int offset, len, naddr, nsize;
+	const uint64_t *regs;
+	uint64_t reg_uart_base;
+
+	uk_pr_info("Serial initializing\n");
+
+	offset = fdt_node_offset_by_compatible(dtb, -1, "ns16550");
```

Comment:
For the [`pl011` driver](https://github.com/unikraft/unikraft/blob/staging/plat/common/arm/pl011.c#L122) the compatible attribute is `arm,pl011`. Should then this be `"arm,ns16550"` instead of just `"ns16550"`?

Reply:
Not needed. Unlike `pl011`, `ns16550` is not IP or Arm Ltd. In general it is a good idea to use the established values in the compatible attribute. From what I see in https://elixir.bootlin.com/linux/v4.9/source/Documentation/devicetree/bindings/serial/8250.txt, there is a long list of possible strings used. I updated this PR to additionally detect `ns16550a`. Unfortunately libfdt does not provide a lookup function against a list, so for now it has to be implemented as a simple conditional. I think we will need that in the future, so I'll make a PR upstream. If it is not accepted I'll add it in our helpers under `plat/drivers/ofw/`.